Frameworks/Foundation/NSOperation.mm

```diff
@@ -1,115 +1,112 @@
-    id dependencies;
-    void (^completionBlock)(void);
+@implementation NSOperation {
+    StrongId<NSMutableArray> _dependencies;
+    void (^_completionBlock)(void);
+    pthread_cond_t _finishCondition;
+    pthread_mutex_t _finishLock;
+    pthread_mutex_t _dependenciesLock;
+    pthread_mutex_t _completionBlockLock;
+}
 
-    pthread_cond_t finishCondition;
-    pthread_mutex_t finishLock;
+@synthesize cancelled = _cancelled;
+@synthesize executing = _executing;
+@synthesize finished = _finished;
 
-    int executing : 1;
-    int cancelled : 1;
-    int finished : 1;
+/**
+ @Status Interoperable
+*/
++ (BOOL)automaticallyNotifiesObserversForKey:(NSString*)key {
+    // This class dispatches its own notifications.
+    return NO;
+}
 
-    NSOperationPriv() {
-        memset(this, 0, sizeof(NSOperationPriv));
-        pthread_cond_init(&finishCondition, 0);
+- (void)observeValueForKeyPath:(NSString*)keyPath ofObject:(id)object change:(NSDictionary*)change context:(void*)context {
+    if ([keyPath isEqualToString:@"isFinished"]) {
+        pthread_mutex_lock(&_finishLock);
+        pthread_cond_broadcast(&_finishCondition);
+        pthread_mutex_unlock(&_finishLock);
+    }
+}
 
+- (id)init {
+    if (self = [super init]) {
+        _dependencies.attach([NSMutableArray new]);
+        pthread_cond_init(&_finishCondition, 0);
         pthread_mutexattr_t attr;
-
         pthread_mutexattr_init(&attr);
         pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
-        pthread_mutex_init(&finishLock, &attr);
+        pthread_mutex_init(&_finishLock, &attr);
         pthread_mutexattr_destroy(&attr);
-    }
+        pthread_mutex_init(&_dependenciesLock, NULL);
+        pthread_mutex_init(&_completionBlockLock, NULL);
 
-    ~NSOperationPriv() {
-        pthread_cond_destroy(&finishCondition);
-        pthread_mutex_destroy(&finishLock);
+        [self addObserver:self forKeyPath:@"isFinished" options:0 context:NULL];
     }
-};
-#else
-struct NSOperationPriv;
-#endif
 
-@interface NSOperation () {
-    struct NSOperationPriv* priv;
+    return self;
 }
-@end
-
-@implementation NSOperation
 
 /**
  @Status Interoperable
 */
-+ (BOOL)automaticallyNotifiesObserversForKey:(NSString*)key {
-    // This class dispatches its own notifications.
-    return NO;
+- (void)addDependency:(id)operation {
+    pthread_mutex_lock(&_dependenciesLock);
+    [_dependencies addObject:operation];
+    pthread_mutex_unlock(&_dependenciesLock);
 }
 
 /**
  @Status Interoperable
 */
-+ (id)allocWithZone:(NSZone*)zone {
-    NSOperation* ret = [super allocWithZone:zone];
-
-    ret->priv = new NSOperationPriv();
-
-    return ret;
+- (void)removeDependency:(NSOperation*)operation {
+    pthread_mutex_lock(&_dependenciesLock);
+    [_dependencies removeObject:operation];
+    pthread_mutex_unlock(&_dependenciesLock);
 }
 
 /**
  @Status Interoperable
 */
-- (NSOperationQueuePriority)queuePriority {
-    return priv->priority;
-}
+- (BOOL)isReady {
```

Comment:
This will need a test as well

---
In reply to: [69355570](https://github.com/Microsoft/WinObjC/pull/602#discussion_r69355570) [](ancestors = 69355570)